### DISCUSSION TITLE
Is it possible to make data object properties read-only?

### DISCUSSION THREAD

**Discussion by jameshulse** [spatie/laravel-data#468] [Q&A]
I am using a data object in an Eloquent cast for a JSON column. When the data object is returned from the database I would prefer if it had public methods to update its state, and allow the application to read its properties but not modify them.

Is this currently possible?

- I tried making the properties private in the constructor, but then Laravel-data couldn’t create the object
- I don’t believe I can mark the whole class as “readonly” because I want to update the object state

Maybe another attribute on the constructor fields like #[Readonly] could use the magic __set() method to catch when a field is being incorrectly set.

**Comment by rubenvanassche** [comment]
Making properties protected is the best solution here I guess. I want to stay away from magic methods as much as possible. I don't think we always want to include protected properties. But if it can be somehow configureable that would be great.

I don't have time to implement such a thing, but I'm always open for well tested and documented PR's!
